conmon-rs/server/src/oom_watcher.rs

```diff
@@ -0,0 +1,270 @@
+        let memory_cgroup_file_oom_path = memory_cgroup_path.join("memory.oom_control");
+        let event_control_path = memory_cgroup_path.join("cgroup.event_control");
+        debug!(
+            "{}: memory_cgroup_file_oom_path {}",
+            pid,
+            memory_cgroup_file_oom_path.display()
+        );
+
+        let oom_cgroup_file = tokio::fs::OpenOptions::new()
+            .write(true)
+            .open(memory_cgroup_file_oom_path)
+            .await
+            .context("opening cgroup file")?;
+        let mut oom_event_fd = EventFd::new(0, false).context("creating eventfd")?;
+
+        let mut event_control = tokio::fs::OpenOptions::new()
+            .write(true)
+            .open(event_control_path)
+            .await
+            .context("opening cgroup file")?;
+        event_control
+            .write_all(
+                format!(
+                    "{} {}",
+                    oom_event_fd.as_raw_fd(),
+                    oom_cgroup_file.as_raw_fd()
+                )
+                .as_bytes(),
+            )
+            .await
+            .context("writing control data")?;
+        event_control.flush().await.context("flush control data")?;
+
+        debug!("{}: successfully setup cgroup v1 oom detection", pid);
+
+        let mut buffer = [0u8; 16];
+        loop {
+            tokio::select! {
+                _ = exit_rx.recv() => {
+                    break;
+                }
+                _ = oom_event_fd.read(&mut buffer) => {
+                    debug!("{}: cgroup v1 oom event", pid);
+                    if let Err(e) = Self::write_oom_files(exit_paths).await {
+                        error!("{}: error writing oom files: {}", pid, e);
+                    } else {
+                        debug!("{}: successfully wrote oom files", pid);
+                    }
+                }
+            }
+        }
+
+        debug!("{}: done watching for ooms", pid);
+        Ok(())
+    }
+
+    fn async_watcher() -> Result<(RecommendedWatcher, Receiver<notify::Result<Event>>)> {
+        let (tx, rx) = channel(1);
+
+        let watcher = RecommendedWatcher::new(move |res: Result<Event, Error>| {
+            futures::executor::block_on(async {
+                tx.send(res).await.unwrap();
+            })
+        })?;
+
+        Ok((watcher, rx))
+    }
+
+    async fn oom_handling_cgroup_v2(
+        exit_rx: &mut BroadcastReceiver<ExitChannelData>,
+        pid: u32,
+        exit_paths: &[PathBuf],
+    ) -> Result<()> {
+        debug!("{}: enabled cgroup v2 oom handling", pid);
+        let subsystem_path = Self::process_cgroup_subsystem_path(pid, true, "memory").await?;
+        let memory_events_file_path = subsystem_path.join("memory.events");
+        let mut last_counter: u64 = 0;
+
+        let (mut watcher, mut rx) = Self::async_watcher()?;
+        debug!(
+            "{}:watching oom path {}",
```

Comment:
nit, no space between pid and message like in others

Reply:
fixed